### PR TITLE
Requestify generic signatures

### PR DESCRIPTION
**Pull request by CodaFi**
This just does the thing that @slavapestov's patch made trivial.  The real work is refining that assertion to read

```cxx
assert(!GenericSigAndBit.getInt() && "Generic signature cannot be changed");
```

so we can catch where we're double-computing signatures (*cough* validateDeclForNameLookup).

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
Jenkins crashed...

@swift-ci please smoke test macOS platform

**Comment by slavapestov** [comment]
This is a great first step, but we also want to refactor the various places under validateDecl() where we call setGenericSignature() -- we ought to build the signature inside the evaluate() method.

File: include/swift/AST/NameLookupRequests.h [hunk 1]
```diff
@@ -273,6 +274,27 @@ class GenericParamListRequest :
   void cacheResult(GenericParamList *value) const;
 };
 
+class GenericSignatureRequest :
```
Comment (slavapestov) [inline include/swift/AST/NameLookupRequests.h]: This should be in TypeCheckRequests since the implementation of evaluate() will use the functions in TypeCheckGeneric.cpp.

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -856,18 +856,21 @@ GenericParamList *GenericContext::getGenericParams() const {
...
-  if (auto PD = dyn_cast<ProtocolDecl>(this)) {
+  if (auto PD = dyn_cast<ProtocolDecl>(GC)) {
     auto self = PD->getSelfInterfaceType()->castTo<GenericTypeParamType>();
     auto req =
         Requirement(RequirementKind::Conformance, self, PD->getDeclaredType());
-    const_cast<GenericContext *>(this)->GenericSig
-      = GenericSignature::get({self}, {req});
-    return GenericSig;
+    return GenericSignature::get({self}, {req});
   }
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: This should be moved to TypeCheckGeneric.cpp, and you can handle NominalTypeDecl, TypeAliasDecl, ExtensionDecl, AbstractFunctionDecl and SubscriptDecl there too by calling into the existing code.

File: include/swift/AST/NameLookupRequests.h (hunk 1 above)
Comment (CodaFi) [reply include/swift/AST/NameLookupRequests.h]: The generic signature is computed by the verifier so clients not linking Sema all fail.

File: include/swift/AST/NameLookupRequests.h [hunk 3]
```diff
@@ -27,6 +27,7 @@ class ClassDecl;
 class DestructorDecl;
 class GenericContext;
 class GenericParamList;
+class GenericSignature;
```
Comment (slavapestov) [inline include/swift/AST/NameLookupRequests.h]: Still needed here?

File: lib/AST/GenericSignatureBuilder.cpp [hunk 4]
```diff
@@ -5296,6 +5296,9 @@ class GenericSignatureBuilder::InferRequirementsWalker : public TypeWalker {
     auto decl = ty->getAnyNominal();
     if (!decl) return Action::Continue;
 
+    if (!decl->hasComputedGenericSignature())
```
Comment (slavapestov) [inline lib/AST/GenericSignatureBuilder.cpp]: Why is this necessary? I would expect that kicking off a request from the GSB is totally the right thing to do

File: lib/Sema/TypeCheckDecl.cpp [hunk 5]
```diff
@@ -3781,15 +3802,15 @@ void TypeChecker::validateDecl(ValueDecl *D) {
     // Check generic parameters, if needed.
     DeclValidationRAII IBV(typeAlias);
 
-    validateGenericTypeSignature(typeAlias);
+    (void)typeAlias->getGenericSignature();
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: This shouldn't be called from validateDecl(). You should force the request in typeCheckDecl() (or rather DeclChecker::visitTypeAliasDecl(), etc, instead)

File: lib/Sema/TypeCheckDecl.cpp [hunk 6]
```diff
@@ -4024,11 +4043,18 @@ void TypeChecker::validateDecl(ValueDecl *D) {
         break;
       }
     }
-
-    validateGenericFuncOrSubscriptSignature(FD, FD, FD);
     
     // We want the function to be available for name lookup as soon
     // as it has a valid interface type.
+    (void)FD->getGenericSignature();
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: Again, let's not force the request here

File: lib/Sema/TypeCheckDecl.cpp [hunk 7]
```diff
@@ -4559,10 +4449,8 @@ void TypeChecker::validateExtension(ExtensionDecl *ext) {
     // Validate the nominal type declaration being extended.
     validateDecl(nominal);
 
-    if (auto *genericParams = ext->getGenericParams()) {
-      auto *sig = checkExtensionGenericParams(*this, ext, genericParams);
-      ext->setGenericSignature(sig);
-    }
+    // Force the generic signature to be computed.
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: I think it should now be possible to remove validateExtension() altogether

File: lib/Sema/TypeCheckProtocolInference.cpp [hunk 8]
```diff
@@ -200,7 +200,7 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
     // The extension may not have a generic signature set up yet, as a
     // recursion breaker, in which case we can't yet confidently reject its
     // witnesses.
-    if (!extension->getGenericSignature())
+    if (!extension->hasComputedGenericSignature())
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocolInference.cpp]: What happens if you just remove this?

File: lib/Sema/TypeCheckGeneric.cpp [hunk 9]
```diff
@@ -649,32 +608,119 @@ void TypeChecker::validateGenericTypeSignature(GenericTypeDecl *typeDecl) {
...
+  // declarations. This is a compile-time optimization since it lets us
+  // avoid the requirements-gathering phase, but it also simplifies that
+  // work for accessors which don't mention the value type in their formal
+  // signatures (like the read and modify coroutines, since yield types
+  // aren't tracked in the AST type yet).
+  if (auto accessor = dyn_cast<AccessorDecl>(GC->getAsDecl())) {
+    if (auto subscr = dyn_cast<SubscriptDecl>(accessor->getStorage())) {
+      return subscr->getGenericSignature();
+    }
+    llvm_unreachable("Wat?");
```
Comment (slavapestov) [inline lib/Sema/TypeCheckGeneric.cpp]: Convert the above dyn_cast to a cast, then remove this unreachable

File: lib/Sema/TypeCheckGeneric.cpp [hunk 10]
```diff
@@ -649,32 +608,119 @@ void TypeChecker::validateGenericTypeSignature(GenericTypeDecl *typeDecl) {
...
+            resultTypeRepr, TypeResolverContext::FunctionResult);
+
+        inferenceSources.emplace_back(resultTypeRepr, resultType);
+      }
+    }
+  } else if (auto *ext = dyn_cast<ExtensionDecl>(GC)) {
+    // Form the interface type of the extension so we can use it as an inference
+    // source.
+    //
+    // FIXME: Push this into the "get interface type" request.
```
Comment (slavapestov) [inline lib/Sema/TypeCheckGeneric.cpp]: Actually I'd like to redo how the inference works here so that formExtensionInterfaceType() is not needed at all.

**Comment by slavapestov** [comment]
Really excited to see this come together!

File: lib/AST/GenericSignatureBuilder.cpp (hunk 4 above)
Comment (CodaFi) [reply lib/AST/GenericSignatureBuilder.cpp]: This is a weird compromise.  The GSB does cycle detection, and so does the request evaluator.  We wind up double-diagnosing because of it in a few tests.

File: lib/Sema/TypeCheckProtocolInference.cpp (hunk 8 above)
Comment (CodaFi) [reply lib/Sema/TypeCheckProtocolInference.cpp]: Asking for the generic signature kicks off type resolution which kicks off protocol witness  resolution which kicks off protocol inference which kicks off this request again. 😢

Comment (slavapestov) [reply lib/Sema/TypeCheckProtocolInference.cpp]: Ah, something to untangle later then. Can you file a radar?

File: lib/AST/GenericSignatureBuilder.cpp (hunk 4 above)
Comment (slavapestov) [reply lib/AST/GenericSignatureBuilder.cpp]: We should clean this up later. Can you file a radar?

**Comment by CodaFi** [comment]
Tried to remove `validateExtension` last night and it led to a bunch of fallout and other cleanup that clutters this PR massively. Going to address your feedback then see if I can  use this to unblock @xymus’ PR because validateDeclForNameLookup stands in our way.

File: lib/AST/GenericSignatureBuilder.cpp (hunk 4 above)
Comment (CodaFi) [reply lib/AST/GenericSignatureBuilder.cpp]: rdar://55263708.  Will attach in a FIXME as well.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
Passed Swift, need LLDB patches.

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
Alright, at least locally this is sufficient hedging to get things to build

Canary first

@swift-ci please smoke test

**Comment by CodaFi** [comment]
Cool, back to LLDB failures.  Off to compat

@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
apple/swift-lldb#1991
@swift-ci please test

**Comment by CodaFi** [comment]
⛵️
